### PR TITLE
Tpetra: Fixing calls to Kokkos::finalize() to match Kokkos 4.0 semantics

### PR DESCRIPTION
**Pull request by csiefer2**
Issue: #11034 , https://github.com/kokkos/kokkos/issues/5476
Tests: HIP (where the issue can be seen with 3.7)
Stakeholder feedback: n/a

### CODE REVIEW

**Review by cgcgcg — DISMISSED** [review]

**Review by ndellingwood — DISMISSED** [review]

**Comment by csiefer2** [comment]
> cuda/11 build and test failures reported here: https://sems-atlassian-son.sandia.gov/jira/servicedesk/customer/portal/7/TRILINOSHD-209 https://sems-atlassian-son.sandia.gov/jira/servicedesk/customer/portal/7/TRILINOSHD-210

@ndellingwood Those will be fixed by https://github.com/trilinos/Trilinos/pull/11043

**Comment by ndellingwood** [comment]
@csiefer2 I'll update, thanks. There are also some other runtime failures (muelu, teko) but I don't think those are related to #11043 right? Only the additive schwartz test?

**Comment by csiefer2** [comment]
@ndellingwood That will fix the compile and disable Ifpack2_AdditiveSchwarz for UVM-off builds.

File: packages/tpetra/core/src/Tpetra_Core.cpp [hunk 1]
```diff
@@ -313,7 +313,7 @@ namespace Tpetra {
 
     // Tpetra should only finalize Kokkos if it initialized Kokkos.
     // See Github Issue #434.
-    if (tpetraInitializedKokkos_) {
+    if (tpetraInitializedKokkos_ && !Kokkos::is_finalized()) {
```
Comment (dalg24) [inline packages/tpetra/core/src/Tpetra_Core.cpp:316]: Why did you change?  Who would call `Kokkos::finalize` if Tpetra initialized Kokkos?

**Review by jhux2 — APPROVED** [review]
oh yeah

**Review by jhux2 — APPROVED** [review]
🐤 lgtm

File: packages/tpetra/core/src/Tpetra_Core.cpp (hunk 1 above)
Comment (dalg24) [reply packages/tpetra/core/src/Tpetra_Core.cpp:316]: ping

Comment (csiefer2) [reply packages/tpetra/core/src/Tpetra_Core.cpp:316]: Hopefully nobody.  But it doesn't hurt to be safe.

Comment (dalg24) [reply packages/tpetra/core/src/Tpetra_Core.cpp:316]: I think this is semantically incorrect.  If you initialized Kokkos you are responsible for finalizing.
If someone else called `finalize` then there is a bug.  The change proposed here hide the bug they don't fix it.

**Comment by ndellingwood** [comment]
cuda-10.1.243 failures are spurious system issues
```
BEFORE: jsrun  '-p' '1' '--rs_per_socket' '4' '/vscratch1/trilinos/jaas/workspace/Trilinos_PR_cuda-10.1.243/pull_request_test/packages/adelus/test/vector_random_fs/Adelus_vector_random_fs.exe' '3501' '1' '1' '1'
AFTER: export TPETRA_ASSUME_CUDA_AWARE_MPI=1; jsrun  '-M -disable_gpu_hooks' '-p' '1' '--rs_per_socket' '4' '/vscratch1/trilinos/jaas/workspace/Trilinos_PR_cuda-10.1.243/pull_request_test/packages/adelus/test/vector_random_fs/Adelus_vector_random_fs.exe' '3501' '1' '1' '1'
Error: error in ptssup_mkcltsock_afunix()
09-26-2022 17:19:39:486 18600 main: Error initializing RM connection. Exiting.
jsrun return value: 1
```

AT removed the RETEST label, I'll re-add to retrigger testing

**Comment by bartlettroscoe** [comment]
Please stop putting `AT: RETEST` on this.  It is not going to pass.

See

* #11072
